### PR TITLE
[Mechanic] Thrust Vectoring

### PR DESCRIPTION
**Pull request by Azure3141**
-----------------------
**Feature:** This PR implements the feature request detailed and discussed in issue #404notfound

## Feature Details
Adds a new attribute under engine flares, `gimbal`, which takes an angle and rotates the engine flare by that amount depending on whether the ship is both thrusting and turning. Works on reverse engines, steering engines, and afterburners as well.

## UI Screenshots
A demonstration of thrust vectoring on the Lampyrid, both with forward and reverse thrusters. (Outdated but basically shows what you would get if you put `gimbal 30` on every engine flare)

https://github.com/endless-sky/endless-sky/assets/42621251/c2e89328-92c2-4dd0-bdd9-4f70a7729a20

## Usage Examples
```
	engine 123 456
		zoom 1
		angle 0
		gimbal 90
		under

```

## Testing Done
Tested vectored thrust with both reverse and forward thrusters to make sure the sprite rotation was behaving

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

**Comment by Hecter94** [comment]
This would look much better if the vectored thrust flare sprite transitioned smoothly between states rather than instantly snapping upon input.

**Comment by Ferociousfeind** [comment]
damn, everyone stealing my complaint

I would _also_ add that the 30 degrees used now seems a little too far, but I haven't seen what half of that rotation looks like to tell whether or not it is better

**Comment by Zitchas** [comment]
Does this automatically generate vectored thrust flares on every thrust flare point?

If so, could we get some way to specify that a flare point should *not* be used for vectored thrust?

For instance, in the case of the lampyrid here, the vectored thrust looks great on the main drive; but is inappropriate on the side thrusters. Having the main drive vector would definitely increase the rotation speed of the ship; but having the side pods vector like that would actually be counterproductive, or more accurately, might make it spin faster, but also push it into a much larger radius turn than it would otherwise.

Anyway, both looks cool and is a neat mechanic to add. I like it!

**Comment by Amazinite** [comment]
I don't think we should be adding more complexity to movement mechanics. As a visual thing of being able to have "thrust" flares that turn when you're turning, this is fine, but as a movement mechanic thing, I'm not a fan.

**Comment by Saugia** [comment]
Agreeing with Derpy - looking at the PR I assumed it was the visuals, which I do like and think would be a neat addition, however, I don't think we need a new mechanic entirely for thrust vectoring, as it doesn't really add anything except yet another type of thrust just for the sake of having another.

**Comment by Ferociousfeind** [comment]
I disagree- even though it seems like a simple change for the sake of change, it's a big important difference which changes the way that you think about movement, and, to be extra snarky it's the type of movement implied by the turning thruster sprites already, which have no counterbalancing thrusters to zero out linear velocity, meaning thrusting and turning are clearly inseparable.

I would love to have it for my aliens which use thrust vectoring _instead of_ dedicated maneuvering thrusters because it's cheaper and easier than having dedicated thrusters just for turning you around (every kilogram counts, pre-repulsor-era)

It just needs AI changes so AI understands how to use it. That's an interesting design problem, because thrust and turn are linked, so you have to solve both of them simultaneously instead of one at a time like we currently do. Though, also, we have less freedom, we're either thrusting and turning, or just thrusting, or waiting for the right moment to start thrusting.

**Review by Amazinite — CHANGES_REQUESTED** [review]
As mentioned, I would prefer if this were a purely visual thing.

File: source/Ship.cpp [hunk 1]
```diff
@@ -4452,12 +4457,17 @@ void Ship::DoEngineVisuals(vector<Visual> &visuals, bool isUsingAfterburner)
 	if(isUsingAfterburner && !Attributes().AfterburnerEffects().empty())
 		for(const EnginePoint &point : enginePoints)
 		{
+			double gimbalDirection = (Commands().Has(Command::FORWARD) || Commands().Has(Command::BACK))
+				* -Commands().Turn();
+			Angle gimbal = Angle(gimbalDirection * point.gimbal.Degrees());
+
 			Point pos = angle.Rotate(point) * Zoom() + position;
+
```
Comment (Amazinite) [inline source/Ship.cpp:4465]:
```suggestion
```

File: source/Ship.cpp [hunk 2]
```diff
@@ -4452,12 +4457,17 @@ void Ship::DoEngineVisuals(vector<Visual> &visuals, bool isUsingAfterburner)
 	if(isUsingAfterburner && !Attributes().AfterburnerEffects().empty())
 		for(const EnginePoint &point : enginePoints)
 		{
+			double gimbalDirection = (Commands().Has(Command::FORWARD) || Commands().Has(Command::BACK))
+				* -Commands().Turn();
```
Comment (Amazinite) [inline source/Ship.cpp:4461]: Move this outside of the for loop but inside the if statement.

**Review by Amazinite — APPROVED** [review]

**Pull request merged**